File: problems/2026/04/2026-04-09-leetcode-101-对称二叉树/SolutionLC101.java
/**
 * LeetCode 101：对称二叉树（递归判断左右子树是否互为镜像）
 */
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class SolutionLC101 {

    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }
        return isMirror(root.left, root.right);
    }

    /**
     * 以 p、q 为根的子树是否互为镜像：根值相等，且 p 左与 q 右镜像、p 右与 q 左镜像。
     */
    private boolean isMirror(TreeNode p, TreeNode q) {
        if (p == null && q == null) {
            return true;
        }
        if (p == null || q == null) {
            return false;
        }
        if (p.val != q.val) {
            return false;
        }
        return isMirror(p.left, q.right) && isMirror(p.right, q.left);
    }

    public static void main(String[] args) {
        SolutionLC101 solution = new SolutionLC101();

        // [1,2,2,3,4,4,3] 对称
        TreeNode sym = new TreeNode(1,
                new TreeNode(2, new TreeNode(3), new TreeNode(4)),
                new TreeNode(2, new TreeNode(4), new TreeNode(3)));
        System.out.println("对称树: " + solution.isSymmetric(sym));

        // [1,2,2,null,3,null,3] 不对称
        TreeNode asym = new TreeNode(1,
                new TreeNode(2, null, new TreeNode(3)),
                new TreeNode(2, null, new TreeNode(3)));
        System.out.println("非对称树: " + solution.isSymmetric(asym));
    }
}

File: problems/2026/04/2026-04-09-leetcode-101-对称二叉树/SolutionLC101.py
from __future__ import annotations

from typing import Optional


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: Optional["TreeNode"] = None,
        right: Optional["TreeNode"] = None,
    ):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    """LeetCode 101：对称二叉树（递归镜像）"""

    def is_symmetric(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True
        return self._is_mirror(root.left, root.right)

    def _is_mirror(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if p is None and q is None:
            return True
        if p is None or q is None:
            return False
        if p.val != q.val:
            return False
        return self._is_mirror(p.left, q.right) and self._is_mirror(p.right, q.left)


if __name__ == "__main__":
    sol = Solution()
    sym = TreeNode(
        1,
        TreeNode(2, TreeNode(3), TreeNode(4)),
        TreeNode(2, TreeNode(4), TreeNode(3)),
    )
    print("对称:", sol.is_symmetric(sym))
    asym = TreeNode(
        1,
        TreeNode(2, None, TreeNode(3)),
        TreeNode(2, None, TreeNode(3)),
    )
    print("非对称:", sol.is_symmetric(asym))
